File: Reverse Parenthesis.py
# Decodes a string by reversing substrings inside parentheses. For each closing parenthesis, the substring inside the matching parentheses is reversed in place.
# Space Complexity: O(n)
# Time Complexity: O(n^2)
def decode(s):
    stack = []

    for char in s:
        if char == ')':
            # pop until matching '('
            temp = []
            while stack and stack[-1] != '(':
                temp.append(stack.pop())
            stack.pop()  # remove '('
            # push reversed substring back to stack
            stack.extend(temp)
        else:
            stack.append(char)

    return ''.join(stack)

File: Reverse Parenthesis.rs
// Time Complexity: O(n^2)
// Space Complexity: O(n)
fn decode(s: &str) -> String {
    let mut stack: Vec<char> = Vec::new();

    for c in s.chars() {
        if c == ')' {
            let mut temp = Vec::new();
            while let Some(&top) = stack.last() {
                if top == '(' {
                    stack.pop(); // remove '('
                    break;
                }
                temp.push(stack.pop().unwrap());
            }
            // push reversed substring back
            stack.extend(temp);
        } else {
            stack.push(c);
        }
    }

    stack.iter().collect()
}

